src/ws.nim

```diff
@@ -0,0 +1,415 @@
+    b1 = 126u8
+  else:
+    b1 = 127u8
+
+  if f.mask:
+    b1 = b1 or (1 shl 7)
+
+  ret.add(uint8 b1)
+
+  # Only need more bytes if data len is 7+16 bits, or 7+64 bits.
+  if f.data.len > 125 and f.data.len <= 0xffff:
+    # Data len is 7+16 bits.
+    var len = f.data.len.uint16
+    ret.add ((len shr 8) and 255).uint8
+    ret.add (len and 255).uint8
+  elif f.data.len > 0xffff:
+    # Data len is 7+64 bits.
+    var len = f.data.len
+    ret.add ((len shr 56) and 255).uint8
+    ret.add ((len shr 48) and 255).uint8
+    ret.add ((len shr 40) and 255).uint8
+    ret.add ((len shr 32) and 255).uint8
+    ret.add ((len shr 24) and 255).uint8
+    ret.add ((len shr 16) and 255).uint8
+    ret.add ((len shr 8) and 255).uint8
+    ret.add (len and 255).uint8
+
+  var data = f.data
+
+  if f.mask:
+    # If we need to mask it generate random mask key and mask the data.
+    for i in 0..<data.len:
+      data[i] = (data[i].uint8 xor f.maskKey[i mod 4].uint8)
+    # Write mask key next.
+    ret.add(f.maskKey[0].uint8)
+    ret.add(f.maskKey[1].uint8)
+    ret.add(f.maskKey[2].uint8)
+    ret.add(f.maskKey[3].uint8)
+
+  # Write the data.
+  ret.add(data)
+  return ret
+
+proc send*(ws: WebSocket, data: seq[byte], opcode = Opcode.Text): Future[
+    void] {.async.} =
+  try:
+    var maskKey: array[4, char]
+    if ws.masked:
+      maskKey = genMaskKey(ws.rng)
+    var frame = encodeFrame((
+      fin: true,
+      rsv1: false,
+      rsv2: false,
+      rsv3: false,
+      opcode: opcode,
+      mask: ws.masked,
+      data: data,
+      maskKey: maskKey
+    ))
+    const maxSize = 1024*1024
+    # Send stuff in 1 megabyte chunks to prevent IOErrors.
+    # This really large packets.
+    var i = 0
+    while i < frame.len:
+      let frameSize = min(frame.len, i + maxSize)
+      let res = await ws.tcpSocket.write(frame[i ..< frameSize])
+      if res != frameSize:
+        raise newException(ValueError, "Error while send websocket frame")
+      i += maxSize
+  except IOError, OSError, ValueError:
+    # Wrap all exceptions in a WebSocketError so its easy to catch
+    raise newException(WebSocketError, "Failed to send data: " &
+        getCurrentExceptionMsg())
+
+proc sendStr*(ws: WebSocket, data: string, opcode = Opcode.Text): Future[
+                void] {.async.} =
+  await send(ws, toBytes(data), opcode)
+
+proc close*(ws: WebSocket) =
+  ## Close the Socket, sends close packet.
+  ws.readyState = Closed
+  proc close() {.async.} =
+    await ws.send(@[], Close)
+    ws.tcpSocket.close()
+  asyncCheck close()
+
+proc receiveFrame(ws: WebSocket): Future[Frame] {.async.} =
+  ## Gets a frame from the WebSocket.
+  ## See https://tools.ietf.org/html/rfc6455#section-5.2
+
+  # Grab the header.
+  var header = newSeq[byte](2)
+  try:
+    await ws.tcpSocket.readExactly(addr header[0], 2)
+  except TransportUseClosedError:
+    ws.readyState = Closed
+    raise newException(WebSocketError, "Socket closed")
+  except CatchableError:
```

Comment:
By catching `CatchableError` you swallowing cancellation exceptions, so its better to avoid such construction or comment about reasons.

Reply:
Done